boards/PDM/Src/Io/Io_Efuse.c

```diff
@@ -0,0 +1,947 @@
+
+    // Read original content of GCR Register
+    RETURN_CODE_IF_EXIT_NOT_OK(Io_Efuse_Aux1Aux2ReadRegister(SI_GCR_ADDR, &reg_val));
+
+    CLEAR_BIT(reg_val, (CSNS1_EN_MASK | CSNS0_EN_MASK));
+    SET_BIT(reg_val, (selection & (CSNS1_EN_MASK | CSNS0_EN_MASK)));
+
+    RETURN_CODE_IF_EXIT_NOT_OK(Io_Efuse_Aux1Aux2WriteRegister(SI_GCR_ADDR, reg_val));
+
+    return EXIT_CODE_OK;
+}
+
+ExitCode Io_Efuse_Aux1Aux2ConfigureEfuse(void)
+{
+    RETURN_CODE_IF_EXIT_NOT_OK(Io_Efuse_Aux1Aux2ExitFailSafeMode());
+
+    // Global config register
+    RETURN_CODE_IF_EXIT_NOT_OK(
+        Io_Efuse_Aux1Aux2WriteRegister(SI_GCR_ADDR, GCR_CONFIG));
+
+    // Channel 0 config registers
+    RETURN_CODE_IF_EXIT_NOT_OK(
+        Io_Efuse_Aux1Aux2WriteRegister(SI_RETRY_0_ADDR, RETRY_CONFIG));
+    RETURN_CODE_IF_EXIT_NOT_OK(
+        Io_Efuse_Aux1Aux2WriteRegister(SI_CONFR_0_ADDR, CONFR_CONFIG));
+    RETURN_CODE_IF_EXIT_NOT_OK(Io_Efuse_Aux1Aux2WriteRegister(
+        SI_OCR_0_ADDR, OCR_LOW_CURRENT_SENSE_CONFIG));
+
+    // Channel 1 config registers
+    RETURN_CODE_IF_EXIT_NOT_OK(
+        Io_Efuse_Aux1Aux2WriteRegister(SI_RETRY_1_ADDR, RETRY_CONFIG));
+    RETURN_CODE_IF_EXIT_NOT_OK(
+        Io_Efuse_Aux1Aux2WriteRegister(SI_CONFR_1_ADDR, CONFR_CONFIG));
+    RETURN_CODE_IF_EXIT_NOT_OK(Io_Efuse_Aux1Aux2WriteRegister(
+        SI_OCR_1_ADDR, OCR_LOW_CURRENT_SENSE_CONFIG));
+
+    return EXIT_CODE_OK;
+}
+
+static ExitCode Io_Efuse_Aux1Aux2ExitFailSafeMode(void)
+{
+    // Set WDIN bit for next write
+    // 1_1_00000_00000_0000
+    aux1_aux2_efuse.wdin_bit_to_set = true;
+
+    RETURN_CODE_IF_EXIT_NOT_OK(
+        Io_Efuse_Aux1Aux2WriteRegister(SI_STATR_0_ADDR, 0x0000));
+    // Disable watchdog
+    RETURN_CODE_IF_EXIT_NOT_OK(
+        Io_Efuse_Aux1Aux2WriteRegister(SI_GCR_ADDR, GCR_CONFIG));
+
+    // Check if the the efuse is still in fail-safe mode
+    if (HAL_GPIO_ReadPin(FSOB_AUX1_AUX2_GPIO_Port, FSOB_AUX1_AUX2_Pin) ==
+        GPIO_PIN_RESET)
+    {
+        return EXIT_CODE_UNIMPLEMENTED;
+    }
+
+    return EXIT_CODE_OK;
+}
+
+static ExitCode Io_Efuse_Aux1Aux2WriteRegister(
+    uint8_t  register_address,
+    uint16_t register_value)
+{
+    return Io_Efuse_WriteRegister(
+        register_address, register_value, CSB_AUX1_AUX2_GPIO_Port,
+        CSB_AUX1_AUX2_Pin, &aux1_aux2_efuse);
+}
+
+static ExitCode Io_Efuse_Aux1Aux2ReadRegister(
+    uint8_t   register_address,
+    uint16_t *register_value)
+{
+    return Io_Efuse_ReadRegister(
+        register_address, register_value, CSB_AUX1_AUX2_GPIO_Port,
+        CSB_AUX1_AUX2_Pin, &aux1_aux2_efuse);
+}
+
+static ExitCode Io_Efuse_WriteRegister(
+    uint8_t               register_address,
+    uint16_t              register_value,
+    GPIO_TypeDef *        chip_select_port,
+    uint16_t              chip_select_pin,
+    struct Efuse_Context *e_fuse)
+{
+    uint16_t serial_input_data = 0x0000U;
+
+    // Place the register address into bits 10->13
+    serial_input_data = (uint16_t)(
```

Comment:
We can use SET_BIT here

Reply:
I think for these using the SET_BIT macro makes it harder to read what is happening. Ken suggested using it for the Watchdog one because it was a simple bitwise operation. It doesn't look great with shifting bits as well

Reply:
Actually why can't we set the serial input data directly here? 

ORing zeroes with   ```((register_address & EFUSE_ADDR_MASK) << EFUSE_ADDR_SHIFT));``` should just be equal to setting the value directly